locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,457 @@
+But because Node has only a few threads, you must structure your application to use them wisely.
+
+Here's a good rule of thumb for keeping your Node server speedy:
+*Node is fast when the work associated with each client at any given time is "small"*.
+
+This applies to callbacks on the Event Loop and tasks on the Worker Pool.
+
+## Why should I avoid blocking the Event Loop and the Worker Pool?
+Node uses a small number of threads to handle many clients.
+In Node there are two types of threads: one Event Loop (aka the main loop, main thread, event thread, etc.), and a pool of `k` Workers in a Worker Pool (aka the threadpool).
+
+If a thread is taking a long time to execute a callback (Event Loop) or a task (Worker), we call it "blocked".
+While a thread is blocked working on behalf of one client, it cannot handle requests from any other clients.
+This provides two motivations for blocking neither the Event Loop nor the Worker Pool:
+
+1. Performance: If you regularly perform heavyweight activity on either type of thread, the *throughput* (requests/second) of your server will suffer.
+2. Security: If it is possible that for certain input one of your threads might block, a malicious client could submit this "evil input", make your threads block, and keep them from working on other clients. This would be a [Denial of Service](https://en.wikipedia.org/wiki/Denial-of-service_attack) attack.
+
+## A quick review of Node
+
+Node uses the Event-Driven Architecture: it has an Event Loop for orchestration and a Worker Pool for expensive tasks.
+
+### What code runs on the Event Loop?
+When they begin, Node applications first complete an initialization phase, `require`'ing modules and registering callbacks for events.
+Node applications then enter the Event Loop, responding to incoming client requests by executing the appropriate callback.
+This callback executes synchronously, and may register asynchronous requests to continue processing after it completes.
+The callbacks for these asynchronous requests will also be executed on the Event Loop.
+
+The Event Loop will also fulfill the non-blocking asynchronous requests made by its callbacks, e.g., network I/O.
+
+In summary, the Event Loop executes the JavaScript callbacks registered for events, and is also responsible for fulfilling non-blocking asynchronous requests like network I/O.
+
+### What code runs on the Worker Pool?
+Node's Worker Pool is implemented in libuv ([docs](http://docs.libuv.org/en/v1.x/threadpool.html)), which exposes a general task submission API.
+
+Node uses the Worker Pool to handle "expensive" tasks.
+This includes I/O for which an operating system does not provide a non-blocking version, as well as particularly CPU-intensive tasks.
+
+These are the Node module APIs that make use of this Worker Pool:
+1. I/O-intensive
```

Comment:
worth mentioning `net` module?

Reply:
It is my understanding that the [`net`](https://nodejs.org/api/net.html) module does *not* use the Worker Pool, thanks to good OS support for non-blocking network I/O. The module docs do not bear the usual caveats about the threadpool, and as far as I can tell from the source there is no call to `uv_queue_work`.

Am I wrong in this? Or are you suggesting I explicitly mention the `net` module as a "safe" option? If so, I attempted to do so in the section "What code runs on the Event Loop?" with the sentence "The Event Loop will also fulfill the non-blocking asynchronous requests made by its callbacks, e.g., network I/O.".

Reply:
correct, net does not use threads